src/Pages/Goals.cshtml

```diff
@@ -1,36 +1,108 @@
-    </div>
-</div>
+<button type="button" class="btn btn-primary text-light mt-3" data-toggle="modal" data-target="#AddGoalModal">
+    Add Goal
+</button>
 
-<div class="container row p-0 mx-auto mb-5">
-    <div class="col-sm border m-3 bg-primary">
-        <a href="#"><h3 class="text-light mt-5 mb-5">Add Goal</h3></a>
+@if (!Model.Goals.Any())
+{
+    <div class=" border container bg-primary mt-5 mb-5">
+        <blockquote class="blockquote text-center">
+            <p class="mb-0 text-light">If you’re bored with life – you don’t get up every morning with a burning desire to do things – you don’t have enough goals.</p>
+            <footer class="blockquote-footer text-dark">Lou Holtz </footer>
+        </blockquote>
     </div>
-    <div class="col-sm border m-3 bg-primary">
-        <a href="#"><h3 class="text-light mt-5 mb-5">Add Goal</h3></a>
-    </div>
-    <div class="col-sm border m-3 bg-primary">
-        <a href="#"><h3 class="text-light mt-5 mb-5">Add Goal</h3></a>
-    </div>
-</div>
+
+}
+else
+{
+    var num = 0;
+    <form method="post">
+        @foreach (var item in Model.Goals)
+        {
+            if (num % 3 == 0)
+            {
+                @Html.Raw("<div class=\"row p-0 mx-auto mt-3\">")
+            }
+            <div class="border m-3 bg-primary col-sm">
+                <button asp-page-handler="Delete" asp-route-id="@item.Id" class="btn btn-primary float-right" type="submit" title="Remove Goal"><i class="far fa-trash-alt text-danger "></i></button>
+                <div class="clearfix"></div>
+                <h3 class="text-light problematic-title" title="Title">@item.Title </h3>
+                <div class="progress mt-4">
+                    @{
+                        var percent = decimal.Round(Model.Savings / item.Amount * 100);
+                        percent = percent > 100 ? 100 : percent;
+                    }
+                    <div class="progress-bar bg-success" role="progressbar" style="width: @percent%;" aria-valuenow="@percent" aria-valuemin="0" aria-valuemax="100" title="Progress">@percent%</div>
+                </div>
+                <h5 class="text-light" title="Saved/Required">@Model.Savings/@item.Amount</h5>
+                @if (percent == 100)
+                {
+                    <button asp-page-handler="Purchased" asp-route-id="@item.Id" asp-route-title="@item.Title" asp-route-amount="@item.Amount" class="btn btn-primary" type="submit" title="Add Goal as expense and remove it"><i class="fas fa-shopping-cart fa-2x text-success"></i></button>
+                }
+                else
+                {
+                    <i class="fas fa-piggy-bank fa-2x text-warning" title="Still saving"></i>
+                }
+
+            </div>
+            num++;
+            if (num % 3 == 0)
+            {
+                @Html.Raw("</div>")
+            }
+        }
+    </form>
+    if (num % 3 != 0)
+    {
+        @Html.Raw("</div>")
+    }
+}
```

Comment:
I really like the system you created for displaying of goals, could you explain a bit how it works?

Reply:
- Basically it goes through all goals, and when number is divisible by 3 i.e. **0**,3,6... (because max 3 elements in one row) it creates new row:
` @Html.Raw("<div class=\"row p-0 mx-auto mt-3\">")` - just a tag with class that puts items into one line.
- After adding goal, if it's number is divisible by 3 it closes the div, since full row is made (pay attention that closing happens after new item added and incremented, and opening before:
`if (num % 3 == 0)
            {
                @Html.Raw("</div>")
            }`
- After going through all items if last row was not closed (because it was not filled to have 3 items, it checks that (number is not divisible by 3 and closes last row):
`if (num % 3 != 0)
    {
        @Html.Raw("</div>")
    }`

_This is the logic of displaying all goals "dinamically", if you have questions about other aspects of this code don't hesitate to ask._